test/external/dataplane/connectivity.go

```diff
@@ -38,179 +38,216 @@ const (
+   ext-app             |    N/A    |       R(*2)       |        R(*2)          |
+   gateway-cluster     |    R(*2)  |       N/A         |        R(*1)          |
+   non-gateway-cluster |    R(*2)  |       R(*1)       |        N/A            |
 
-		// Get handle for existing docker.
-		docker := framework.New(extAppName)
+   Legend: N: Not reachable, R: Reachable (source IP isn't globalIP),
+           S: Source IP is global IP (and reachable)
 
-		// Get IPs to use later.
-		podIP := np.Pod.Status.PodIP
-		svcIP := svc.Spec.ClusterIP
-		dockerIP := docker.GetIP(extNetName)
+   (*1) Not covered in this test, but covered in normal connectivity tests.
+   (*2) Pod w/ hostnetwork isn't reachable.
+*/
 
-		By(fmt.Sprintf("Sending an http request from external app %q to the service %q in the cluster %q",
-			dockerIP, svcIP, clusterName))
+var _ = Describe("[external-dataplane] Connectivity", func() {
+	f := framework.NewFramework("ext-dataplane")
 
-		command := []string{"curl", "-m", "10", fmt.Sprintf("%s:%d/%s%s", svcIP, 80, f.Namespace, clusterName)}
-		_, _ = docker.RunCommand(command...)
+	var toEndpointType tcp.EndpointType
+	var networking framework.NetworkingType
+	var cluster framework.ClusterIndex
+	var err error
 
-		By("Verifying the pod received the request")
+	verifyInteraction := func(clusterScheduling framework.NetworkPodScheduling) {
+		It("should be able to connect from an external app to a pod in a cluster", func() {
+			if framework.TestContext.GlobalnetEnabled {
+				framework.Skipf("Globalnet is enabled, skipping the test...")
+				return
+			}
 
-		podLog := np.GetLog()
+			testExternalConnectivity(testParams{
+				Framework:         f,
+				ToEndpointType:    toEndpointType,
+				Networking:        networking,
+				Cluster:           cluster,
+				ClusterScheduling: clusterScheduling,
+			})
+		})
+	}
 
-		if clusterName == externalClusterName {
-			Expect(podLog).To(MatchRegexp(".*GET /%s%s .*", f.Namespace, clusterName))
-		} else {
-			Expect(podLog).To(MatchRegexp("%s .*GET /%s%s .*", dockerIP, f.Namespace, clusterName))
-		}
+	When("connected from a gateway-cluster", func() {
+		BeforeEach(func() {
+			cluster, err = getGatewayClusterIndex(framework.TestContext.ClusterIDs)
+			Expect(err).NotTo(HaveOccurred())
+		})
 
-		By(fmt.Sprintf("Sending an http request from the test pod %q %q in cluster %q to the external app %q",
-			np.Pod.Name, podIP, clusterName, dockerIP))
+		When("a pod connects via TCP to a remote pod", func() {
+			BeforeEach(func() {
+				toEndpointType = tcp.PodIP
+				networking = framework.PodNetworking
+			})
 
-		cmd := []string{"curl", "-m", "10", fmt.Sprintf("%s:%d/%s%s", dockerIP, 80, f.Namespace, clusterName)}
-		_, _ = np.RunCommand(cmd)
+			// Access from a pod on NonGatewayNodes to external apps is not supported for an external-app-connected cluster
 
-		By("Verifying that external app received request")
+			When("the pod is on a gateway", func() {
```

Comment:
Please use "Context" where necessary to frame an appropriate sentence for each of the test-case.

Reply:
I agree that nesting to many `When` makes the log hard to read, but it would be easier to maintain the code. I meant to write the codes in the same way to [tcp_gn_pod_connectivity.go](https://github.com/submariner-io/submariner/blob/devel/test/e2e/dataplane/tcp_gn_pod_connectivity.go#L65). I've added `PWhen` to show that there are another pending cases. If it is better to remove pending cases and reduce the nest, please let me know (I don't have strong objection to it).